vulture_os/toolkit/api_parser/carbon_black/carbon_black.py

```diff
@@ -0,0 +1,180 @@
+
+            return True
+
+        except Exception as err:
+            raise CarbonBlackAPIError(err)
+
+    def __execute_query(self, method, url, query, timeout=10):
+
+        self._connect()
+
+        if method == "GET":
+            response = self.session.get(
+                url,
+                params=query,
+                headers=self.HEADERS,
+                timeout=timeout,
+                proxies=self.proxies
+            )
+        elif method == "POST":
+            response = self.session.post(
+                url,
+                json=query,
+                headers=self.HEADERS,
+                timeout=timeout,
+                proxies=self.proxies
+            )
+        else:
+            raise CarbonBlackAPIError(f"Request method unrecognized : {method}")
+
+        if response.status_code != 200:
+            raise CarbonBlackAPIError(f"Error at CarbonBlack API Call URL: {url} Code: {response.status_code} Content: {response.content}")
+
+        return response.json()
+
+    def test(self):
+        try:
+            logs = self.get_logs(since=(datetime.utcnow()-timedelta(days=10)).isoformat()+"Z")
+
+            return {
+                "status": True,
+                "data": [self.format_log(log) for log in logs['results']]
+            }
+        except Exception as e:
+            logger.exception(e)
+            return {
+                "status": False,
+                "error": str(e)
+            }
+
+    def get_logs(self, cursor=0, since=None):
+        alert_url = self.ALERTS.format(self.carbon_black_host, self.carbon_black_orgkey)
+
+        # Format timestamp for query, API wants a Z at the end
+        if isinstance(since, datetime):
+            since = since.isoformat()
+        since = since.replace("+00:00", "Z")
+        if since[-1] != "Z":
+            since += "Z"
+
+        query =  {
+            'criteria': {
+                'category': ['THREAT'],
+                'group_results': True,
+                'last_update_time': {
+                    'start': since,
+                    'end': datetime.now().strftime("%Y-%m-%dT%H:%M:%S.000Z")
+                }
+            },
+            'rows': 9999,
+            'start': cursor
+        }
+
+        return self.__execute_query("POST", alert_url, query)
+
+    def format_log(self, log):
+        log['observer_name'] = self.carbon_black_host.replace("https://", "")
+        log['url'] = self.carbon_black_host
+        return json.dumps(log)
+
+
+    def execute(self):
+
+        since = self.last_api_call or (datetime.utcnow() - timedelta(hours=24))
+
+        available = 0
+        found = 0
+        while found <= available:
+
+            response = self.get_logs(found, since)
+
+            # Downloading may take some while, so refresh token in Redis
+            self.update_lock()
+
+            logs = response['results']
+            found = int(response['num_found']) + 1
+            available = int(response['num_found'])
```

Comment:
Soomething wrong here, 'found' and 'available' will never be useful

Reply:
Yes you're right. I fixed this loop according to documentation.
num_found can be higher than num_available.